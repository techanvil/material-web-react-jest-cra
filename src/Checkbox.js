/**
 * External dependencies
 */
import "@material/web/checkbox/checkbox";
import PropTypes from "prop-types";
import React from "react";

/**
 * WordPress dependencies
 */
//  import { useEffect, useRef, useState } from '@wordpress/element';

import { useEffect, useRef, useState } from "react";

const log = (...args) => console.__proto__.log.call(console, ...args);

export default function Checkbox({
  checked,
  disabled,
  name,
  value,
  onChange,
  onKeyDown,
}) {
  log("Checkbox render. checked:", checked);
  const ref = useRef(null);

  // Use this state in a key prop on the web component to force a rerender.
  // This is a workaround for the fact the web component doesn't update visually when the checked property is changed programmatically. Hopefully this will be fixed in a future update to the web component.
  // It has a side effect of unfocusing the checkbox when changed via keyboard input which needs to be addressed.
  const [index, setIndex] = useState(0);

  useEffect(() => {
    const { current } = ref;

    const click = (event) => {
      log("click", event);

      // Prevent default and manually dispatch a change event, in order to retain checkbox state and use as a controlled input.
      event.preventDefault();

      if (disabled) {
        return;
      }

      current.checked = current.checked ? undefined : true;
      // current.checked = ! current.checked;

      current.dispatchEvent(
        new Event("change", {
          bubbles: true,
          composed: true,
        })
      );
    };

    const change = (event) => {
      log("change", event);

      // Preventing default behaviour has no effect here.
      // event.preventDefault();

      onChange?.(event);

      setIndex(index + 1);
    };

    // Keydown events work fine without any special logic.
    const keydown = (event) => {
      log("keydown", event);

      onKeyDown?.(event);
    };

    // The click event is triggered by both mouse and keyboard entry (space key) on Chrome, need to confirm other browsers.
    current?.addEventListener("click", click);
    current?.addEventListener("change", change);
    current?.addEventListener("keydown", keydown);

    // The 'action' event is dispatched by the ActionElement base class of the Material checkbox. See ActionElement source for details.
    // current?.addEventListener( 'action', cancelEvent );

    return () => {
      current?.removeEventListener("click", click);
      current?.removeEventListener("change", change);
      current?.removeEventListener("keydown", keydown);
    };
  }, [checked, disabled, index, onChange, onKeyDown]);

  // TODO: Change theme colour to match expected ?

  return (
    <md-checkbox
      key={`checkbox-${name}-${index}`}
      ref={ref}
      role="checkbox"
      // Lit boolean attributes treat anything non-null|undefined as true. Coerce to undefined if false.
      // See https://lit.dev/docs/v1/components/properties/#attributes
      checked={checked || undefined}
      disabled={disabled || undefined}
      name={name}
      value={value}
    ></md-checkbox>
  );
}

/**
 * Props determined by examining the Material 3 web component's source code.
 *
 * See @property definitions and event handlers in the source to identify which props are needed.
 */

Checkbox.propTypes = {
  // Fundamental checkbox attributes.
  checked: PropTypes.bool,
  disabled: PropTypes.bool,
  name: PropTypes.string,
  value: PropTypes.string,

  // Event handlers.
  onChange: PropTypes.func,
  onKeyDown: PropTypes.func,

  // Accessibility attributes.
  // ariaLabel: PropTypes.string,
  // ariaLabelledBy: PropTypes.string,
  // ariaDescribedBy: PropTypes.string,
  // reducedTouchTarget: PropTypes.bool,
};
