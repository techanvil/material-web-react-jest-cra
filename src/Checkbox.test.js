import React from "react";
import { useState } from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import Checkbox from "./Checkbox";

// TODO: Worth a look - https://www.npmjs.com/package/shadow-dom-testing-library

const log = (...args) => console.__proto__.log.call(console, ...args);

describe("Checkbox", () => {
  it("should render the checkbox", () => {
    const { container } = render(<Checkbox />);
    expect(container).toMatchSnapshot();
  });

  describe("controlled input behaviour", () => {
    it("should correctly invoke onChange and retain its checked state when clicked", () => {
      const onChange = jest.fn();
      const { container } = render(<Checkbox onChange={onChange} />);

      expect(container).toMatchSnapshot();

      // Confirm the checkbox is not checked.
      expect(screen.getByRole("checkbox")).not.toHaveAttribute("checked");

      fireEvent.click(screen.getByRole("checkbox"));

      expect(container).toMatchSnapshot();

      // Confirm the click resulted in a change event with target.checked: true.
      expect(onChange).toHaveBeenCalledTimes(1);
      const event = onChange.mock.calls[0][0];
      expect(event.target.checked).toBe(true);

      // Confirm the checkbox remains unchecked as its state is controlled.
      expect(screen.getByRole("checkbox")).not.toHaveAttribute("checked");
    });

    it("should allow updating of the checked state", () => {
      function CheckableCheckbox({ onChange }) {
        const [checked, setChecked] = useState(false);

        return (
          <Checkbox
            checked={checked}
            onChange={(event) => {
              onChange(event);
              setChecked(event.target.checked);
            }}
          />
        );
      }

      const onChange = jest.fn();
      const { container } = render(<CheckableCheckbox onChange={onChange} />);

      expect(container).toMatchSnapshot();

      // Confirm the checkbox is not checked.
      expect(screen.getByRole("checkbox")).not.toHaveAttribute("checked");

      fireEvent.click(screen.getByRole("checkbox"));

      expect(container).toMatchSnapshot();

      // Confirm the click resulted in a change event with target.checked: true.
      expect(onChange).toHaveBeenCalledTimes(1);
      let event = onChange.mock.calls[0][0];
      expect(event.target.checked).toBe(true);

      // Confirm the checkbox is now checked
      expect(screen.getByRole("checkbox")).toHaveAttribute("checked");

      // Click again to uncheck
      onChange.mockClear();
      fireEvent.click(screen.getByRole("checkbox"));

      expect(container).toMatchSnapshot();

      // Confirm the click resulted in a change event with target.checked: undefined.
      expect(onChange).toHaveBeenCalledTimes(1);
      event = onChange.mock.calls[0][0];
      expect(event.target).toHaveProperty("checked", undefined);

      // Confirm the checkbox is now unchecked
      expect(screen.getByRole("checkbox")).not.toHaveAttribute("checked");
    });
  });
});
