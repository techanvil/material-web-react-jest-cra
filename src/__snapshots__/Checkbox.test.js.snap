// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox should render the checkbox 1`] = `
<div>
  <md-checkbox />
</div>
`;
